src/expression.rs

```diff
@@ -81,14 +84,18 @@ impl Expression {
     /// Consume the expression, simplifying it as much as possible using the values provided in the environment.
     /// If variables are used in the expression which are not present in the environment, evaluation stops there,
     /// returning the possibly-simplified expression.
-    pub fn evaluate(self, environment: &EvaluationEnvironment) -> Self {
+    pub fn evaluate(
+        self,
+        environment: &EvaluationEnvironment,
+        patch_values: Option<&HashMap<&str, Vec<f64>>>,
```

Comment:
I don't normally like optional-containers-of-possibly-empty-things (does `None` mean something different from `empty`?), but I think this is about as clean as it can be without Rust allowing default keyword arguments; passing an empty `HashMap::new()` instead of `None` wouldn't exactly be zero-cost.

Reply:
That is why, it's just because `None` is easier to type and also zero-cost.

Reply:
I defo get it! I’m just still scarred from working with a Scala `Option[Seq[Option[String]]]` at a previous job 😭

Reply:
Do Scala Options implement from iterator? Meaning, can you you do like in Rust and easily turn `Vec<Option<Thing>>` into `Option<Vec<Thing>>`?

Between that and `map`/`and_then` I rarely have issues making Optional code both readable and efficient 😁.

In any event, the refactor of `evaluate`/`simplify` eliminated this `Option`.

Reply:
It does have a `flatMap`, which means you can use them in `for` comprehensions (FYI `for` in Scala is a monad comprehension, _not_ a for loop!) but the standard library lacks the traverse/sequence that [`Futures`](https://www.scala-lang.org/api/2.13.5/scala/concurrent/Future$.html) have. Luckily, the [`cats`](https://typelevel.org/cats/) library adds that good stuff in, but that leads to completely different programming idioms.

<img width="515" alt="Screen Shot 2021-08-19 at 9 43 50 PM" src="https://user-images.githubusercontent.com/884550/130165943-51548c65-0f78-4a75-907b-b66b49802169.png">